This PR implements setter and getters methods for the DH parameters

juanjqo opened the pull request:
Hi @dqrobotics/developers, 

This PR implements setter and getter methods for the DH parameters in the `DQ_SerialManipulatorDH` and `DQ_SerialManipulatorMDH` classes, which are already available in the [MATLAB version](https://github.com/dqrobotics/matlab/pull/106). This PR aims to complete one of the [pending tasks for the 23.04 release](https://github.com/dqrobotics/developers-playground/discussions/1).

<img src="https://github.com/user-attachments/assets/06edf57b-1096-47be-8065-a38cfe6f4f8d" alt="drawing" width="400"/>


#### List of Modifications:

- New enum class: `DQ_ParameterDH` ~in the `DQ_SerialManipulator` class~.
```cpp
namespace DQ_robotics
{
enum class DQ_ParameterDH
{
    THETA,
    D,
    A,
    ALPHA
};
}
```
- The `DQ_SerialManipulatorDH` and `DQ_SerialManipulatorMDH` classes now have the methods 
```cpp
    VectorXd get_parameters(const DQ_ParameterDH& parameter_type) const;
    double   get_parameter(const DQ_ParameterDH& parameter_type,
                           const int& to_ith_link) const;
    void set_parameters(const DQ_ParameterDH& parameter_type,
                        const VectorXd& vector_parameters);
    void set_parameter(const DQ_ParameterDH& parameter_type,
                       const int& to_ith_link,
                       const double& parameter);
```


### Usage:

```cpp
    auto robot = FrankaEmikaPandaRobot::kinematics();
    robot.set_parameter(DQ_ParameterDH::D, 2, 0.523);

    // Update: Using char/string style is now allowed, as in Matlab
    robot.set_parameters("THETA", VectorXd::Zero(7));
```
### Example:

I proposed an example in https://github.com/dqrobotics/cpp-examples/pull/20.

Kind regards, 

Juancho

mmmarinho commented:
Hi @juanjqo. Thanks for this. I noticed you changed this to draft so I hope this comment helps, I don't mean to imply I think this is the finished product. My memory is fuzzy but I remember `DQ_ParameterDH` being its own class, not necessarily a nested class of `DQ_SerialManipulator`.

I think this architecture you proposed would be great (and cleaner) in `cpp` and `Python`. My concern is that we'll soon get some issues raised that languages differ, because `matlab` does not have the trailing `DQ_SerialManipulator` which is effectively just behaving as a namespace.

juanjqo commented:
> Hi @juanjqo. Thanks for this. I noticed you changed this to draft so I hope this comment helps, I don't mean to imply I think this is the finished product. My memory is fuzzy but I remember `DQ_ParameterDH` being its own class, not necessarily a nested class of `DQ_SerialManipulator`.
> 
> I think this architecture you proposed would be great (and cleaner) in `cpp` and `Python`. My concern is that we'll soon get some issues raised that languages differ, because `matlab` does not have the trailing `DQ_SerialManipulator` which is effectively just behaving as a namespace.

Hi @mmmarinho, thanks for your feedback. I am going to implement the required modifications taking into account your comments.

mmmarinho commented:
PS: before you throw a keyboard at me, here’s an example we can start from:
https://stackoverflow.com/questions/21295935/can-a-c-enum-class-have-methods

juanjqo commented:
Hi @mmmarinho, I updated the `DQ_ParameterDH` class to support the char/string style in the setter and getter methods. I updated the [example](https://github.com/dqrobotics/cpp-examples/pull/20) to test it.  

```cpp
    auto robot = FrankaEmikaPandaRobot::kinematics();
    robot.set_parameter("THETA", 2, 0.01);
    std::cout<<robot.get_parameter("THETA", 2)<<std::endl;
```

PS: There is a good chance that I misunderstood your instructions. Please tell me if my modifications are close to what you were thinking.

Kind regards, 

Juancho

mmmarinho commented:
@juanjqo seems good to me!

Could I please ask you to document in doxygen style all the methods, including the new constructions etc? You’ve done it for most but maybe not for all. Apologies if my phone isn’t showing it correctly.

juanjqo commented:
> @juanjqo seems good to me!
> 
> Could I please ask you to document in doxygen style all the methods, including the new constructions etc? You’ve done it for most but maybe not for all. Apologies if my phone isn’t showing it correctly.

@mmmarinho I added the documentation for the constructors. =)
